File: commandapi-platforms/commandapi-bukkit/commandapi-bukkit-test/commandapi-bukkit-test-tests/src/test/java/dev/jorel/commandapi/test/CommandAPIServerMock.java
package dev.jorel.commandapi.test;

import static org.junit.jupiter.api.Assertions.fail;
import static org.mockito.ArgumentMatchers.isA;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.stream.Stream;

import be.seeseemelk.mockbukkit.plugin.PluginManagerMock;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.Keyed;
import org.bukkit.NamespacedKey;
import org.bukkit.Registry;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.advancement.Advancement;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.inventory.ItemFactory;
import org.bukkit.plugin.messaging.Messenger;
import org.bukkit.plugin.messaging.StandardMessenger;
import org.bukkit.permissions.Permissible;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.ParseResults;
import com.mojang.brigadier.context.StringRange;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.suggestion.Suggestion;
import com.mojang.brigadier.suggestion.Suggestions;

import be.seeseemelk.mockbukkit.ServerMock;
import be.seeseemelk.mockbukkit.WorldMock;
import be.seeseemelk.mockbukkit.entity.PlayerMock;
import dev.jorel.commandapi.Brigadier;
import org.mockito.Mockito;

public class CommandAPIServerMock extends ServerMock {
	public CommandAPIServerMock() {
		// The functionality of these methods is not important, but they shouldn't throw an `UnimplementedOperationException`
		//  These methods are called indirectly when `CommandNamespaceTests#enableWithNamespaces` constructs a `PermissibleBase`
		PluginManagerMock pluginManagerSpy = Mockito.spy(this.getPluginManager());
		Mockito.doNothing().when(pluginManagerSpy).unsubscribeFromDefaultPerms(isA(Boolean.class), isA(Permissible.class));
		Mockito.doNothing().when(pluginManagerSpy).subscribeToDefaultPerms(isA(Boolean.class), isA(Permissible.class));
		MockPlatform.setField(ServerMock.class, "pluginManager", this, pluginManagerSpy);
	}

	@SuppressWarnings("unchecked")
	public boolean dispatchThrowableCommand(CommandSender sender, String commandLine) throws CommandSyntaxException {
		String[] commands = commandLine.split(" ");
		String commandLabel = commands[0];
		Command command = getCommandMap().getCommand(commandLabel);

		if (command != null) {
			return super.dispatchCommand(sender, commandLine);
		} else {
			@SuppressWarnings("rawtypes")
			CommandDispatcher dispatcher = Brigadier.getCommandDispatcher();
			Object css = Brigadier.getBrigadierSourceFromCommandSender(sender);
			return dispatcher.execute(commandLine, css) != 0;
		}
	}

	@Override
	public boolean dispatchCommand(CommandSender sender, String commandLine) {
		try {
			return dispatchThrowableCommand(sender, commandLine);
		} catch (CommandSyntaxException e) {
			fail("Command '/" + commandLine + "' failed. If you expected this to fail, use dispatchThrowableCommand() instead.", e);
			return false;
		}
	}

	public int dispatchBrigadierCommand(CommandSender sender, String commandLine) {
		try {
			return dispatchThrowableBrigadierCommand(sender, commandLine);
		} catch (CommandSyntaxException e) {
			fail("Command '/" + commandLine + "' failed. If you expected this to fail, use dispatchThrowableBrigadierCommand() instead.", e);
			return 0;
		}
	}

	@SuppressWarnings("unchecked")
	public int dispatchThrowableBrigadierCommand(CommandSender sender, String commandLine) throws CommandSyntaxException {
		@SuppressWarnings("rawtypes")
		CommandDispatcher dispatcher = Brigadier.getCommandDispatcher();
		Object css = Brigadier.getBrigadierSourceFromCommandSender(sender);
		return dispatcher.execute(commandLine, css);
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	public List<String> getSuggestions(CommandSender sender, String commandLine) {
		CommandDispatcher dispatcher = Brigadier.getCommandDispatcher();
		Object css = Brigadier.getBrigadierSourceFromCommandSender(sender);
		ParseResults parseResults = dispatcher.parse(commandLine, css);
		Suggestions suggestions = null;
		try {
			suggestions = (Suggestions) dispatcher.getCompletionSuggestions(parseResults).get();
		} catch (InterruptedException | ExecutionException e) {
			e.printStackTrace();
			suggestions = new Suggestions(StringRange.at(0), new ArrayList<>()); // Empty suggestions
		}

		List<String> suggestionsAsStrings = new ArrayList<>();
		for (Suggestion suggestion : suggestions.getList()) {
			suggestionsAsStrings.add(suggestion.getText());
		}

		return suggestionsAsStrings;
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	public List<Suggestion> getSuggestionsWithTooltips(CommandSender sender, String commandLine) {
		CommandDispatcher dispatcher = Brigadier.getCommandDispatcher();
		Object css = Brigadier.getBrigadierSourceFromCommandSender(sender);
		ParseResults parseResults = dispatcher.parse(commandLine, css);
		Suggestions suggestions = null;
		try {
			suggestions = (Suggestions) dispatcher.getCompletionSuggestions(parseResults).get();
		} catch (InterruptedException | ExecutionException e) {
			e.printStackTrace();
			suggestions = new Suggestions(StringRange.at(0), new ArrayList<>()); // Empty suggestions
		}

		List<Suggestion> suggestionsList = new ArrayList<>();
		for (Suggestion suggestion : suggestions.getList()) {
			suggestionsList.add(suggestion);
		}

		return suggestionsList;
	}

//	@Override
	public boolean shouldSendChatPreviews() {
		return true;
	}
	
	// Registries

//	@Override
	public <T extends Keyed> @Nullable Registry<T> getRegistry(@NotNull Class<T> tClass) {
		return MockPlatform.getInstance().getRegistry(tClass);
//		if (tClass.equals(Enchantment.class)) {
//			return new Registry() {
//				@Nullable
//				public T get(@NotNull NamespacedKey var1) {
//					System.out.println("Accessing " + tClass + ":" + var1);
//					return null;
//				}
//				
//				@NotNull
//				public Stream<T> stream() {
//					List<T> list = List.of();
//					return list.stream();
//				}
//
//				public Iterator<T> iterator() {
//					List<T> list = List.of();
//					return list.iterator();
//				}
//			};
//		} else {
//			return null;
//		}
	}
	
	// World mocking
	
	static class CustomWorldMock extends WorldMock {
		
		@Override
		public boolean equals(Object obj) {
			if(obj == null) {
				return false;
			} else if(obj instanceof World target) {
				return this.getUID().equals(target.getUID());
			} else {
				return false; // I have no idea what this is
			}
		}

	}
	
	@Override
	public WorldMock addSimpleWorld(String name) {
		WorldMock world = new CustomWorldMock();
		world.setName(name);
		super.addWorld(world);
		return world;
	}
	
	// TODO: Commenting this out for now because we've not sorted out 1.20.2
	// support with MockBukkit, this almost certainly will require all sorts of
	// hoops to jump through because apparently MockBukkit may have changed the
	// method signature for this and we REALLY REALLY don't want to deal with
	// that right now
//	@Override
//	public ItemFactory getItemFactory() {
//		// Thanks MockBukkit, but we REALLY need to access
//		// the raw CraftItemMeta objects for the ItemStackArgument <3
//		return MockPlatform.getInstance().getItemFactory();
//	}

	// 1.16 and 1.17 MockServers do not implement this method, but other versions do
	//  Easiest to just always override this method
	//  This is copied from MockBukkit-v1.18
	private final StandardMessenger messenger = new StandardMessenger();
	@Override
	public @NotNull Messenger getMessenger() {
		return messenger;
	}

	/**
	 * Creates a new Bukkit {@link Player}. Unlike {@link PlayerMock}, this uses Mockito to mock the CraftPlayer class,
	 * which allows the returned object to pass through VanillaCommandWrapper#getListener without error.
	 *
	 * @return A new {@link Player}.
	 */
	public Player setupMockedCraftPlayer() {
		return setupMockedCraftPlayer("defaultName");
	}

	/**
	 * Creates a new Bukkit {@link Player}. Unlike {@link PlayerMock}, this uses Mockito to mock the CraftPlayer class,
	 * which allows the returned object to pass through VanillaCommandWrapper#getListener without error.
	 *
	 * @param name The name for the player
	 * @return A new {@link Player}.
	 */
	public Player setupMockedCraftPlayer(String name) {
		return MockPlatform.getInstance().setupMockedCraftPlayer(name);
	}

	// Advancements
	
	List<Advancement> advancements = new ArrayList<>();
	
	public void addAdvancement(NamespacedKey key) {
		advancements.add(MockPlatform.getInstance().addAdvancement(key));
	}
	
	public void addAdvancements(Collection<NamespacedKey> key) {
		key.forEach(this::addAdvancement);
	}
	
	@Override
	public Iterator<Advancement> advancementIterator() {
		return advancements.iterator();
	}
}

File: commandapi-platforms/commandapi-bukkit/commandapi-bukkit-test/commandapi-bukkit-kotlin-test/src/test/kotlin/dev/jorel/commandapi/test/CommandAPIServerMock.kt
package dev.jorel.commandapi.test

import be.seeseemelk.mockbukkit.ServerMock
import be.seeseemelk.mockbukkit.entity.PlayerMock
import com.mojang.brigadier.context.StringRange
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.suggestion.Suggestions
import dev.jorel.commandapi.Brigadier
import org.bukkit.Keyed
import org.bukkit.Registry
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.scoreboard.Criteria
import org.junit.jupiter.api.Assertions
import java.util.concurrent.ExecutionException

class CommandAPIServerMock : ServerMock() {
	@Throws(CommandSyntaxException::class)
	fun dispatchThrowableCommand(sender: CommandSender, commandLine: String): Boolean {
		val commands = commandLine.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
		val commandLabel = commands[0]
		val command = commandMap.getCommand(commandLabel)
		return if (command != null) {
			super.dispatchCommand(sender, commandLine)
		} else {
			val dispatcher = Brigadier.getCommandDispatcher()
			val css = Brigadier.getBrigadierSourceFromCommandSender(sender)
			dispatcher.execute(commandLine, css) != 0
		}
	}

	fun dispatchBrigadierCommand(sender: CommandSender, commandLine: String): Int {
		return try {
			val dispatcher = Brigadier.getCommandDispatcher()
			val css = Brigadier.getBrigadierSourceFromCommandSender(sender)
			dispatcher.execute(commandLine, css)
		} catch (e: CommandSyntaxException) {
			Assertions.fail<Any>("Command '/$commandLine' failed. If you expected this to fail, use dispatchThrowableCommand() instead.", e)
			0
		}
	}

	override fun dispatchCommand(sender: CommandSender, commandLine: String): Boolean {
		return try {
			dispatchThrowableCommand(sender, commandLine)
		} catch (e: CommandSyntaxException) {
			Assertions.fail<Any>("Command '/$commandLine' failed. If you expected this to fail, use dispatchThrowableCommand() instead.", e)
			false
		}
	}

	fun getSuggestions(sender: CommandSender, commandLine: String?): List<String> {
		val dispatcher = Brigadier.getCommandDispatcher()
		val css = Brigadier.getBrigadierSourceFromCommandSender(sender)
		val parseResults = dispatcher.parse(commandLine, css)
		var suggestions: Suggestions? = null
		suggestions = try {
			dispatcher.getCompletionSuggestions(parseResults).get() as Suggestions
		} catch (e: InterruptedException) {
			e.printStackTrace()
			Suggestions(StringRange.at(0), ArrayList()) // Empty suggestions
		} catch (e: ExecutionException) {
			e.printStackTrace()
			Suggestions(StringRange.at(0), ArrayList())
		}
		val suggestionsAsStrings: MutableList<String> = ArrayList()
		for (suggestion in suggestions!!.list) {
			suggestionsAsStrings.add(suggestion.text)
		}
		return suggestionsAsStrings
	}

	override fun addPlayer(): PlayerMock {
		val player = super.addPlayer()
		MockPlatform.getInstance<Any>().wrapPlayerMockIntoCraftPlayer(player)
		return player
	}

	override fun addPlayer(name: String): PlayerMock {
		val player = super.addPlayer(name)
		MockPlatform.getInstance<Any>().wrapPlayerMockIntoCraftPlayer(player)
		return player
	}

	/**
	 * Creates a new Bukkit [Player]. Unlike [PlayerMock], this uses Mockito to mock the CraftPlayer class,
	 * which allows the returned object to pass through VanillaCommandWrapper#getListener without error.
	 *
	 * @return A new [Player] with a randome name.
	 */
	fun addCraftPlayer(): Player {
		return MockPlatform.getInstance<Any>().wrapPlayerMockIntoCraftPlayer(super.addPlayer())
	}

	/**
	 * Creates a new Bukkit [Player]. Unlike [PlayerMock], this uses Mockito to mock the CraftPlayer class,
	 * which allows the returned object to pass through VanillaCommandWrapper#getListener without error.
	 *
	 * @param name The name for the player
	 * @return A new [Player].
	 */
	fun addCraftPlayer(name: String?): Player {
		return MockPlatform.getInstance<Any>().wrapPlayerMockIntoCraftPlayer(super.addPlayer(name))
	}

	override fun isTickingWorlds(): Boolean {
		TODO("Not yet implemented")
	}

	override fun shouldSendChatPreviews(): Boolean {
		return true
	}

	override fun isEnforcingSecureProfiles(): Boolean {
		TODO("Not yet implemented")
	}

	override fun getMaxChainedNeighborUpdates(): Int {
		TODO("Not yet implemented")
	}

	override fun getScoreboardCriteria(name: String): Criteria {
		TODO("Not yet implemented")
	}

	override fun <T : Keyed?> getRegistry(tClass: Class<T>): Registry<T>? {
		TODO("Not yet implemented")
	}

}